staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+// Watcher is an example of watching on resource create/update/delete events
+type Watcher struct {
+	clientset     *kubernetes.Clientset
+	exampleClient *rest.RESTClient
+	exampleScheme *runtime.Scheme
+}
+
+// Run starts an Example resource watcher
+func (w *Watcher) Run(ctx context.Context) error {
+	fmt.Printf("Watch Example objects\n")
+
+	// Watch Example objects
+	handler := ExampleEventHandler{}
+	exampleInformer, err := watchExamples(ctx, w.exampleClient, w.exampleScheme, &handler)
+	if err != nil {
+		fmt.Printf("Failed to register watch for Example resource: %v\n", err)
+		return err
+	}
+
+	fmt.Printf("WaitForCacheSync\n")
+	// We must wait for exampleInformer to populate its cache to avoid reading from an empty cache
+	// in case of resource-generated evxents.
+	if !cache.WaitForCacheSync(ctx.Done(), exampleInformer.HasSynced) {
+		return errors.New("Wait for Example Informer was cancelled")
+	}
+
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+func watchExamples(ctx context.Context, exampleClient cache.Getter, exampleScheme *runtime.Scheme, handler cache.ResourceEventHandler) (cache.Controller, error) {
+	parameterCodec := runtime.NewParameterCodec(exampleScheme)
+
+	source := newListWatchFromClient(
+		exampleClient,
+		ExampleResourcePath,
+		api.NamespaceAll,
+		fields.Everything(),
+		parameterCodec)
+
+	store, controller := cache.NewInformer(
+		source,
+
+		// The object type.
+		&Example{},
+
+		// resyncPeriod
+		// Every resyncPeriod, all resources in the cache will retrigger events.
+		// Set to 0 to disable the resync.
+		0,
```

Comment:
A resync event is going to be handled by OnAdd or OnUpdate, so it will be confusing. Plus in most cases a controller don't need to do resync. Hence, I prefer setting it to 0.